src/vs/workbench/contrib/debug/browser/debugHover.ts

```diff
@@ -178,30 +184,37 @@ export class DebugHoverWidget implements IContentWidget {
 		return this.domNode;
 	}
 
-	async showAt(position: Position, focus: boolean): Promise<void> {
+	async showAt(position: Position, focus: boolean): Promise<void | ShowDebugHoverResult> {
 		this.showCancellationSource?.cancel();
 		const cancellationSource = this.showCancellationSource = new CancellationTokenSource();
 		const session = this.debugService.getViewModel().focusedSession;
 
 		if (!session || !this.editor.hasModel()) {
 			this.hide();
-			return;
+			return ShowDebugHoverResult.NOT_AVAILABLE;
 		}
 
 		const result = await this.debugHoverComputer.compute(position, cancellationSource.token);
 		if (this.isVisible() && !result.rangeChanged) {
-			return;
+			return ShowDebugHoverResult.NOT_CHANGED;
 		}
 
-		if (!result.range || cancellationSource.token.isCancellationRequested) {
+		if (!result.range) {
```

Comment:
Now you've removed this check of the cancel token. We should _always_ check the cancellation token after doing an async operation. Otherwise we might do unnecessary work or make side-effects after the operation has already been cancelled.

Reply:
Ah apologies for that. I did not realise that. I now get it why the token is being passed around async functions.

So here is what I think could be the flow of statements

1. After computing the result , firstly if there was a cancellation requested then we return `ShowDebugHoverResult.CANCELLED`
2. Else, if there is no range associated with the result , then we return `ShowDebugHoverResult.NOT_AVAILABLE` 
3. Else, if the result is visible and the range hasn't changed since the previous show then we return `ShowDebugHoverResult.NOT_CHANGED`
4. After evaluating the expression, if there is a cancellation requested then we return `ShowDebugHoverResult.CANCELLED`
5. Finally when there is no expression available , we return `ShowDebugHoverResult.NOT_AVAILABLE`.

![Screenshot 2023-03-28 at 20 28 01](https://user-images.githubusercontent.com/8697234/228333406-d8784d73-68bb-4a5d-b466-764aeb633538.png)

Reply:
Yeah, I think that is correct